src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -96,52 +96,64 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     //update the client CA list 
                     UpdateCAListFromRootStore(innerContext);
                 }
-
-                if (sslAuthenticationOptions.ApplicationProtocols != null)
+                GCHandle alpnHandle = default;
+                try
                 {
-                    if (sslAuthenticationOptions.IsServer)
+                    if (sslAuthenticationOptions.ApplicationProtocols != null)
                     {
-                        byte[] protos = Interop.Ssl.ConvertAlpnProtocolListToByteArray(sslAuthenticationOptions.ApplicationProtocols);
-                        sslAuthenticationOptions.AlpnProtocolsHandle = GCHandle.Alloc(protos, GCHandleType.Pinned);
-                        Interop.Ssl.SslCtxSetAlpnSelectCb(innerContext, s_alpnServerCallback, GCHandle.ToIntPtr(sslAuthenticationOptions.AlpnProtocolsHandle));
-                    }
-                    else
-                    {
-                        if (Interop.Ssl.SslCtxSetAlpnProtos(innerContext, sslAuthenticationOptions.ApplicationProtocols) != 0)
+                        if (sslAuthenticationOptions.IsServer)
                         {
-                            throw CreateSslException(SR.net_alpn_config_failed);
+                            alpnHandle = GCHandle.Alloc(sslAuthenticationOptions.ApplicationProtocols);
```

Comment:
Is this list guaranteed free from mutation during the relevant period?

Reply:
Nope, no guarantee, but is this an issue? Worst case it will just fail the ALPN negotiation. There is no unsafe code operating on the list so its protected by the usual managed runtime properties.

Now if you think that the list should be immutable... maybe,  it's something I argued for during the initial API review, but it was pointed out that this is inconsistent with .net API behaviour, its Caveat Emptor, as long as there is no possibility of memory corruption, modifying something you passed into AuthenticateAsync before it returns is a bug in user code.

Reply:
I had to re-read the callback like 4 times to decide that none of the pointer math was made scarier by this list changing.  But I agree that with the current code it isn't a memory issue.